src/api/callback.cc

```diff
@@ -102,6 +102,7 @@ void InternalCallbackScope::Close() {
   if (!env_->can_call_into_js()) return;
   if (!tick_info->has_tick_scheduled()) {
     MicrotasksScope::PerformCheckpoint(env_->isolate());
+    env_->isolate()->ClearKeptObjects();  // WeakRef
```

Comment:
my intention is for it to be called whenever the js stack empties.

Reply:
Right, but this is inside the conditional here?

Reply:
Sorry, I'm not entirely sure what the consequences of that are. I was just trying to match it up with where we run microtasks, since microtasks have the same invariant.

Reply:
> I was just trying to match it up with where we run microtasks, since microtasks have the same invariant.

We do also run the microtask queue from JS when there are items in the nextTick queue at this point – this is why we skip this call here for that, to avoid redundant work.

We don’t call `ClearKeptObjects` from JS at that point, though, so that’s a case where the call would be missing. (As I understand it, that would also be invalid.)

What I would suggest is to add
`OnScopeLeave cleanup([&]() { env_->isolate->ClearKeptObjects(); });`
near the beginning of this function; in that case, it will always be called.

Reply:
I'm not sure of how this should be translated in node implementation, but the expectation is that ClearKeptObjects would be called after current synchronous code execution has been completed, including any pending microtasks.
In HTML this translates to the microtask checkpoint, after the microstask queue has been drained: https://github.com/whatwg/html/pull/4571